src/pass/zero_elimination.cc

```diff
@@ -0,0 +1,1719 @@
+    res.push_back(GE::make(v->var, v->dom->min));
+    res.push_back(LT::make(v->var, v->dom->min + v->dom->extent));
+  }
+  return res;
+}
+
+// Convert an array of itervars to a map from vars to ranges
+Map<Var, Range> IterVarsToMap(const Array<IterVar>& itervars) {
+  Map<Var, Range> res;
+  for (const IterVar& v : itervars) {
+    res.Set(v->var, v->dom);
+  }
+  return res;
+}
+
+// Convert an array of itervars to an array of vars
+Array<Var> IterVarsToVars(const Array<IterVar>& itervars) {
+  Array<Var> res;
+  for (const IterVar& v : itervars) {
+    res.push_back(v->var);
+  }
+  return res;
+}
+
+// Given a map from vars to ranges create an array of itervars
+Array<IterVar> IterVarsFromMap(const Array<Var>& vars, const Map<Var, Range>& vranges,
+                               IterVarType iter_type = kDataPar, std::string thread_tag = "") {
+  Array<IterVar> res;
+  for (const Var& v : vars) {
+    CHECK(vranges.count(v)) << "A range for the variable " << v
+      << " was not provided in map " << vranges;
+    res.push_back(IterVarNode::make(vranges[v], v, iter_type, thread_tag));
+  }
+  return res;
+}
+
+// Return true if this combiner is just a sum.
+bool IsSumCombiner(const CommReducer& combiner) {
+  if (combiner->result.size() != 1) {
+    return false;
+  }
+
+  if (!is_const_value(SuperSimplify(combiner->identity_element[0]), 0)) {
+    return false;
+  }
+
+  return is_const_value(SuperSimplify(combiner->result[0] -
+                                      (combiner->lhs[0] + combiner->rhs[0])),
+                        0);
+}
+
+// Return true if zero may be factored out of a reduction with this combiner.
+bool CanFactorZeroFromCombiner(const CommReducer& combiner, int value_index) {
+  if (!is_const_value(combiner->identity_element[value_index], 0)) {
+    return false;
+  }
+
+  Expr zero = make_zero(combiner->result[value_index].type());
+  Expr in = Substitute(combiner->result[value_index],
+                       {{combiner->lhs[value_index], zero},
+                        {combiner->rhs[value_index], zero}});
+  in = SuperSimplify(in);
+
+  return is_const_value(in, 0);
+}
+
+Expr InlineThisCall(const Expr& expr) {
+  if (const Call* op = expr.as<Call>()) {
+    if (op->call_type == Call::CallType::Halide) {
+      if (const ComputeOpNode* op_comp = op->func.as<ComputeOpNode>()) {
+        Array<Var> tensor_axes;
+        for (const auto& var : op_comp->axis) {
+          tensor_axes.push_back(var->var);
+        }
+
+        Stmt inlined = Inline(Evaluate::make(expr), op->func, tensor_axes,
+                              op_comp->body[op->value_index]);
+        if (const ir::Evaluate* ev = inlined.as<ir::Evaluate>()) {
+          // If it is a reduction, clone it
+          return CloneReduction(ev->value);
+        }
+      }
+    }
+  }
+
+  return expr;
+}
+
+Tensor InlineTailCall(const Tensor& tensor) {
+  return op::TransformBody(tensor, InlineThisCall);
+}
+
+class InlineTensorsMutator : public IRMutator {
```

Comment:
Add class comments please.

Reply:
Turned out there is a bug in this class; fixed.